File: app/src/main/java/de/ka/chappted/commons/ContextHelper.java
package de.ka.chappted.commons;

import android.app.Activity;
import android.content.Context;
import android.support.annotation.Nullable;

import java.lang.ref.WeakReference;

/**
 * A helper for context classes.
 *
 * Created by Thomas Hofmann on 06.01.18.
 */
public class ContextHelper {

    private static ContextHelper sHelper;

    private WeakReference<Activity> mActivityReference;

    private WeakReference<Context> mApplicationContext;

    private static ContextHelper get() {
        if (sHelper == null) {
            sHelper = new ContextHelper();
        }
        return sHelper;
    }

    public static void init(Context applicationContext) {
        get().mApplicationContext = new WeakReference<>(applicationContext);
    }

    private ContextHelper() {
    }


    public static void onResume(Activity activity) {
        get().mActivityReference = new WeakReference<>(activity);
    }

    @Nullable
    public static Activity getActivityReference() {

        if (get().mActivityReference == null) {
            return null;
        }

        return get().mActivityReference.get();
    }

    @Nullable
    public static Context getAppReference() {

        if (get().mApplicationContext == null) {
            return null;
        }

        return get().mApplicationContext.get();
    }


}

File: app/src/main/java/de/ka/chappted/utils/ContextHelper.kt
package de.ka.chappted.utils

import android.app.Activity
import android.content.Context

import java.lang.ref.WeakReference

/**
 * A helper for context classes.
 *
 * Created by Thomas Hofmann on 06.01.18.
 */
class ContextHelper private constructor() {

    private var activityReference: WeakReference<Activity?>? = null
    private var applicationContext: WeakReference<Context?>? = null

    private object Holder {
        val INSTANCE = ContextHelper()
    }

    companion object {

        val instance: ContextHelper by lazy { Holder.INSTANCE }

        fun init(applicationContext: Context?) {
            instance.applicationContext = WeakReference(applicationContext)
        }

        fun onResume(activity: Activity?) {
            instance.activityReference = WeakReference(activity)
        }

        /**
         * Retrieves the activity reference, if possible. This might not always be the case!
         * instance
         */
        val activityReference: Activity?
            get() = if (instance.activityReference == null) {
                null
            } else instance.activityReference!!.get()

        /**
         * Retrieves the app context reference, if possible. This might not always be the case!
         * Should not be used if it is mandatory to retrieve a context in all cases.
         */
        val appReference: Context?
            get() = if (instance.applicationContext == null) {
                null
            } else instance.applicationContext!!.get()
    }


}
